src/wp-includes/fonts/class-wp-font-collection.php

```diff
@@ -164,14 +188,24 @@ private function load_from_url( $url ) {
 			}
 
 			// Make sure the data is valid before storing it in a transient.
-			$data = $this->sanitize_and_validate_data( $data );
+			$data = $this->sanitize_and_validate_data( $data, array( 'font_families' ) );
 			if ( is_wp_error( $data ) ) {
 				return $data;
 			}
 
 			set_site_transient( $transient_key, $data, DAY_IN_SECONDS );
 		}
 
+		$data['name'] = $this->data['name'];
+
+		if ( isset( $this->data['description'] ) ) {
+			$data['description'] = $this->data['description'];
+		}
+
+		if ( isset( $this->data['categories'] ) ) {
+			$data['categories'] = $this->data['categories'];
+		}
+
```

Comment:
This snippet is repeated in the `load_from_file` method. That seems unnecessary. Could we avoid that?

Reply:
I don‘t think it‘s worth introducing an extra method for this, a little duplication is totally fine and less code than a new method.

Reply:
There's no need to add a new method. The repeated snippet is in `load_from_file` and `load_from_url`. Both functions are only called from `load_from_json` and return `$data`.
We could avoid repetition by removing it from `load_from_file` and `load_from_url` and adding it in `load_from_json`. Something like this:

```php
$data = $url ? $this->load_from_url( $url ) : $this->load_from_file( $file );

$data['name'] = $this->data['name'];

if ( isset( $this->data['description'] ) ) {
	$data['description'] = $this->data['description'];
}

if ( isset( $this->data['categories'] ) ) {
	$data['categories'] = $this->data['categories'];
}

return $data;
```

Reply:
Apart from that, which can be considered a nit; I'm not sure we want to override the PHP-defined properties with the data coming from the JSON file if we use that function signature. If we will be overriding the PHP-defined properties like 'name', `description`, and `categories` if they are defined also in the JSON file, maybe the signature of the function should be different make that evident, as @creativecoder proposed in this comment: https://github.com/WordPress/wordpress-develop/pull/6130#issuecomment-1951681223

Reply:
I think we have two alternatives:
1. If we think that in the future, we won't need to add data properties to the JSON file, we should:
     - stick to the [current function signature](https://github.com/WordPress/wordpress-develop/pull/6130#issuecomment-1952681380)
     - JSON files should contain ONLY a list of font families.
     - No other data that `font_families` is loaded from JSON files (we need to remove the current behavior of loading `name`, `description` and `categories` from the JSON file.

2. If we think that in the future we may need to add data properties to the JSON file, we should:
     - Set the path to the JSON file as [a new parameter](https://github.com/WordPress/wordpress-develop/pull/6130#issuecomment-1951681223) and not as the `font_families` property of `$args`.
     -  Load `name`, `description` and `categories` from the JSON file (if they are not defined in PHP $args).
     
     
I can see benefits in both approaches, so I don't have a strong preference, but if I need to choose, I would lean on option 1. I consider option 1 less flexible but clearer and simpler, while 2 is the opposite. I think the new properties we would need to add to font collections (i.e., https://github.com/WordPress/gutenberg/issues/58774) would be better to add them as new properties of the `$args` parameter.

more points of view are welcome @swissspidy @youknowriad @creativecoder

Reply:
Gotcha. For now, I moved the logic to the other method now to reduce duplication & made sure only the `font_families` keys is used. Makes it easier to reason about.

For flexibility & consistency reasons I really believe a single param like (1) makes most sense. Font families is the most dynamic data of the whole thing and needs to be fetched remotely, everything else (such as license information) can be defined in PHP.

----

Aside:

> https://github.com/WordPress/gutenberg/issues/58774

<rant> Surprised stuff like this is raised so late for something that has been that long in the making. It's pretty obvious that this kind of logic needs to be collection agnostic...  </rant>

Reply:
> Gotcha. For now, I moved the logic to the other method now to reduce duplication & made sure only the font_families keys is used. Makes it easier to reason about.

Cool :+1: this is testing well on my end.

> For flexibility & consistency reasons I really believe a single param like (1) makes most sense. Font families is the most dynamic data of the whole thing and needs to be fetched remotely, everything else (such as license information) can be defined in PHP.

I agree, let's go with that.


> Aside: https://github.com/WordPress/gutenberg/issues/58774. Surprised stuff like this is raised so late for something that has been that long in the making. It's pretty obvious that this kind of logic needs to be collection agnostic...

It should be simple to add. Let's merge this and iterate on that.

Reply:
Follow up for https://core.trac.wordpress.org/ticket/60593 /  https://github.com/WordPress/gutenberg/issues/58774 submitted here: https://github.com/WordPress/wordpress-develop/pull/6153